ouroboros-consensus-cardano/tools/db-converter/Main.hs

```diff
@@ -0,0 +1,326 @@
+  deriving (Show, Typeable)
+
+instance Exception ValidationError
+
+main :: IO ()
+main = do
+  (cmd :: Args Unwrapped) <- unwrapRecord "Byron DB converter"
+  case cmd of
+    Convert {epochDir, dbDir, epochSlots} -> do
+      dbDir' <- parseAbsDir =<< canonicalizePath dbDir
+      (_, files) <- listDir =<< parseAbsDir =<< canonicalizePath epochDir
+      let epochFiles = filter (\f -> fileExtension f == ".epoch") files
+      putStrLn $ "Writing to " ++ show dbDir
+      for_ (sort epochFiles) $ \f -> do
+        putStrLn $ "Converting file " ++ show f
+        convertChunkFile (EpochSlots epochSlots) f dbDir'
+    Validate { dbDir, blockType } ->
+      case blockType of
+        Just Byron   -> validate cmd (Proxy @ ByronBlock)
+        Just Shelley -> validate cmd (Proxy @ (ShelleyBlock TPraosStandardCrypto))
+        Just Cardano -> validate cmd (Proxy @ (CardanoBlock TPraosStandardCrypto))
+        Nothing -> do
+          -- check the dbmarker of the db if the block type is not specified.
+          protocolMagicId <- readDBMarker dbDir
+          case unProtocolMagicId protocolMagicId of
+            764824073  -> validate cmd (Proxy @ByronBlock)
+            1097911063 -> validate cmd (Proxy @ByronBlock)
+            42         -> validate cmd (Proxy @(ShelleyBlock TPraosStandardCrypto))
+            _          -> error $ "unsupported protocolMagicId: " ++ show protocolMagicId
+
+convertChunkFile
+  :: EpochSlots
+  -> Path Abs File -- ^ Input
+  -> Path Abs Dir  -- ^ Ouput directory
+  -> IO (Either CC.ParseError ())
+convertChunkFile es inFile outDir = do
+    createDirIfMissing True dbDir
+    -- Old filename format is XXXXX.epoch, new is XXXXX.chunk
+    outFileName <- parseRelFile (toFilePath (filename inFile))
+    outFile <- (dbDir </> outFileName) -<.> "chunk"
+    IO.withFile (toFilePath outFile) IO.WriteMode $ \h ->
+      runResourceT $ runExceptT $ S.mapM_ (liftIO . BS.hPut h) . S.map encode $ inStream
+  where
+    inStream = CC.parseEpochFileWithBoundary es (toFilePath inFile)
+    dbDir = outDir </> [reldir|immutable|]
+    encode =
+        CB.serializeEncoding'
+      . Byron.encodeByronBlock
+      . Byron.mkByronBlock es
+
+readDBMarker :: FilePath -> IO ProtocolMagicId
+readDBMarker dbPath = do
+    bs <- BS.readFile markerPath
+    protocolMagicId <- runExceptT $ dbMarkerParse markerPath bs
+    return $ fromRight
+      (error "failed to parse protocolMagicId from db Marker file")
+      protocolMagicId
+  where
+    markerPath :: String
+    markerPath = dbPath FilePath.</> Text.unpack dbMarkerFile
+
+validate :: forall blk. (HasProtocolInfo blk, Node.RunNode blk, Show (Header blk)) => Args Unwrapped -> Proxy blk ->  IO ()
+validate Validate {..} _ = do
+  dbDir' <- parseAbsDir =<< canonicalizePath dbDir
+  protocolInfo <- mkProtocolInfo @blk configFile genesisHash requiresNetworkMagic
+  validateChainDb dbDir' protocolInfo onlyImmDB verbose
+validate _ _ = error "impossible"
```

Comment:
Hmm, this makes me think that we should be splitting this up in `db-converter` (leave it in Byron) and `db-validator` (here in Cardano). Yeah, we should do that.